### PR TITLE
Add new loading features

### PR DESCRIPTION
**Pull request by CheeryLee**
**Purpose of the pull request**  
Feature request

**Description about the pull request**  
Following in the footsteps of https://github.com/nutti/Screencast-Keys/issues/25 . I did some changes to make screencast working in two cases:
1) automatically load at startup if the appropriate setting is on;
2) don't turn off screencast if new scene was loaded in the curent session

### CODE REVIEW

**Comment by nutti** [comment]
@CheeryLee 

Thanks for the PR and challenging the difficult feature.
I tried your PR, and I found the below issue.

* When I executed the **File** > **New** > **General**, Screencast Key will be disabled.

**Comment by CheeryLee** [comment]
@nutti whoops, my bad. Made a small fix.

**Comment by nutti** [comment]
@CheeryLee 

There are still errors.

```
2.93\scripts\addons_contrib\screencast_keys\__init__.py", line 119, in get_func
    return ops.SK_OT_ScreencastKeys.is_running()
AttributeError: type object 'SK_OT_ScreencastKeys' has no attribute 'is_running'
```

Also, `is_running` should be in `SK_OT_ScreencastKeys` and global variable is bad way for Blender add-on.

**Comment by CheeryLee** [comment]
Oh, my bad again, there should be `ops.is_running()` or `ops.running` instead of `ops.SK_OT_ScreencastKeys.is_running()`.
Currently this is the one way I see to determine is screencast working or no. I don't think Blender can give any guarantee that operator instance won't be reloaded in a new scene.

**Comment by nutti** [comment]
Thanks. Now, I confirmed it works correctly.

I will check your code in detail from now.

File: src/screencast_keys/__init__.py [hunk 1]
```diff
@@ -115,7 +116,7 @@ def call_silently(fn, *args):
 
 def register_screencast_keys_enable_property():
     def get_func(self):
-        return ops.SK_OT_ScreencastKeys.is_running()
+        return ops.is_running()
```
Comment (nutti) [inline src/screencast_keys/__init__.py:119]: As I mentioned the comment, is it possible to move to the classmethod of `SK_OT_ScreencastKeys`?

File: src/screencast_keys/ops.py [hunk 2]
```diff
@@ -523,17 +537,14 @@ class SK_OT_ScreencastKeys(bpy.types.Operator):
     # TODO: Clear when this model is finished.
     area_spaces = collections.defaultdict(set)
 
-    # Check if this operator is running.
-    # TODO: We can check it with the valid of event handler.
-    running = False
+    # Should screencast be restarted
+    restart: bpy.props.BoolProperty(
+        default=False
+    )
```
Comment (nutti) [inline src/screencast_keys/ops.py:543]:
```suggestion
    # Should screencast be restarted
    restart = bpy.props.BoolProperty(
        default=False
    )
```

And, please add `@compat.make_annotations` under `@BlClassRegistry()` which enables the add-on compatible on Blender 2.7X.

File: src/screencast_keys/ops.py [hunk 3]
```diff
@@ -1371,15 +1382,31 @@ def event_timer_remove(cls, context):
                 wm.event_timer_remove(cls.timers[key])
         cls.timers.clear()
 
-    def invoke(self, context, event):
-        cls = self.__class__
-        prefs = compat.get_user_preferences(context).addons[__package__].preferences
+    @classmethod
+    def start(cls, self, context, event, prefs, is_start):
```
Comment (nutti) [inline src/screencast_keys/ops.py:1386]: I think `is_start` argument is not good for understanding this classmethod.
Could you separate `start` method and `stop` method whether `is_start = True` or `is_start = False`?

File: src/screencast_keys/ops.py [hunk 4]
```diff
@@ -1371,15 +1382,31 @@ def event_timer_remove(cls, context):
...
+        if is_start:
+            if is_running():
+                return
```
Comment (nutti) [inline src/screencast_keys/ops.py:1389]: I think is the impossible case, so return error code and check the value in the caller functions.

File: src/screencast_keys/preferences.py [hunk 5]
```diff
@@ -254,6 +254,12 @@ class SK_Preferences(bpy.types.AddonPreferences):
         default=True,
     )
 
+    load_on_startup = bpy.props.BoolProperty(
+        name="Load On Startup",
+        description="""(Experimental) Automatically enable screencast when blender is starting up""",
```
Comment (nutti) [inline src/screencast_keys/preferences.py:259]: Should this option experimental status?
If so, what is your concern?

File: src/screencast_keys/preferences.py [hunk 6]
```diff
@@ -402,8 +408,9 @@ def draw(self, context):
 
             layout.label(text="Experimental:")
             col = layout.column()
-            col.prop(self, "get_event_aggressively")
-            col.enabled = False if c_structures.NOT_SUPPORTED else True
+            col.prop(self, "load_on_startup")
+            if c_structures.NOT_SUPPORTED == False:
+                col.prop(self, "get_event_aggressively")
```
Comment (nutti) [inline src/screencast_keys/preferences.py:413]: Please remain the **Get Event Aggressively** visible if the option is not supported.

File: src/screencast_keys/ops.py [hunk 7]
```diff
@@ -1507,3 +1539,23 @@ def invoke(self, context, event):
...
+@BlClassRegistry()
+class SK_OT_WaitForInitialize(bpy.types.Operator):
+    bl_idname = "wm.sk_wait_for_initialize"
+    bl_label = "Screencast Keys Wait For Initialize"
+
+    initialization_handler = None
+
+    def execute(self, context):
+        cls = self.__class__
+        cls.initialization_handler = bpy.types.SpaceView3D.draw_handler_add(cls.intialization_callback, (None, None), 'WINDOW', 'POST_PIXEL')
```
Comment (nutti) [inline src/screencast_keys/ops.py:1553]: [Question]
Does calling `bpy.ops.wm.sk_screencast_keys` work in this place?

File: src/screencast_keys/ops.py [hunk 8]
```diff
@@ -1507,3 +1539,23 @@ def invoke(self, context, event):
...
+        print(cls.initialization_handler)
```
Comment (nutti) [inline src/screencast_keys/ops.py:1554]: Please use `debug_print` like [this](https://github.com/nutti/Screencast-Keys/blob/9bf75d109cb3b6c8974adf37a022718e5f1ab846/src/screencast_keys/ops.py#L1130) to output debug print.

File: src/screencast_keys/ops.py [hunk 9]
```diff
@@ -1507,3 +1539,23 @@ def invoke(self, context, event):
         self.draw_handler_add(context)
         context.window_manager.modal_handler_add(self)
         return {'RUNNING_MODAL'}
+
+
+@BlClassRegistry()
+class SK_OT_WaitForInitialize(bpy.types.Operator):
```
Comment (nutti) [inline src/screencast_keys/ops.py:1545]: I think `SK_OT_WaitBlenderInitializedAndStartScreencastKeys` is better name to understand this operator.

File: src/screencast_keys/ops.py [hunk 10]
```diff
@@ -1507,3 +1539,23 @@ def invoke(self, context, event):
...
+    bl_idname = "wm.sk_wait_for_initialize"
+    bl_label = "Screencast Keys Wait For Initialize"
```
Comment (nutti) [inline src/screencast_keys/ops.py:1547]:
```suggestion
    bl_label = "Wait For Blender Initialized And Start Screencast Keys"
```

File: src/screencast_keys/ops.py [hunk 11]
```diff
@@ -1507,3 +1539,23 @@ def invoke(self, context, event):
         self.draw_handler_add(context)
         context.window_manager.modal_handler_add(self)
         return {'RUNNING_MODAL'}
+
+
+@BlClassRegistry()
+class SK_OT_WaitForInitialize(bpy.types.Operator):
+    bl_idname = "wm.sk_wait_for_initialize"
```
Comment (nutti) [inline src/screencast_keys/ops.py:1546]:
```suggestion
    bl_idname = "wm.sk_wait_blender_initialized_and_start_screencast_keys"
```

File: src/screencast_keys/ops.py [hunk 12]
```diff
@@ -1507,3 +1539,23 @@ def invoke(self, context, event):
...
+        cls = self.__class__
+        cls.initialization_handler = bpy.types.SpaceView3D.draw_handler_add(cls.intialization_callback, (None, None), 'WINDOW', 'POST_PIXEL')
+        print(cls.initialization_handler)
+        return {'FINISHED'}
+
+    @classmethod
+    def intialization_callback(cls, self, cotext):
+        if bpy.context.area is not None:
+            bpy.ops.wm.sk_screencast_keys('INVOKE_REGION_WIN', restart=True)
+            bpy.types.SpaceView3D.draw_handler_remove(cls.initialization_handler, 'WINDOW')
```
Comment (nutti) [inline src/screencast_keys/ops.py:1561]: Could you add an empty line.

**Review by nutti — CHANGES_REQUESTED** [review]
Thanks for your great work @CheeryLee !

Before merging this PR, could you check my comments?

File: src/screencast_keys/ops.py (hunk 9 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1545]: Yes

File: src/screencast_keys/__init__.py [hunk 13]
```diff
@@ -57,19 +57,20 @@
 
 
 addon_keymaps = []
+startup = True
```
[the remark this answers is not in the document]
Comment (CheeryLee) [reply src/screencast_keys/__init__.py:60]: No.
`bpy.types.Scene` instance would be overwritten on scene reload. This variable is a marker for determine is Blender just started or this is a next scene load. It's not used anywhere but here.

File: src/screencast_keys/__init__.py (hunk 1 above)
Comment (CheeryLee) [reply src/screencast_keys/__init__.py:119]: Ok. moved it back to operator. Seems like it's working.

File: src/screencast_keys/ops.py (hunk 3 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1386]: Good advice

File: src/screencast_keys/ops.py (hunk 4 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1389]: It's a preventive way to not run unexpected code. But if we have not found any problems, I should remove it.

File: src/screencast_keys/preferences.py (hunk 5 above)
Comment (CheeryLee) [reply src/screencast_keys/preferences.py:259]: I added it here because I don't know your pipeline about introducing new features. If you think it must not be there, I'll move it to the common section.

File: src/screencast_keys/ops.py (hunk 7 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1553]: No. This is the main reason why my changes were created. If there will be direct call of `bpy.ops.wm.sk_screencast_keys`, it won't have any changes beetwen the same call from `load_post_handler`, just a bit deeper. The information about `bpy.context.area` is not loaded at that moment, therefore it must be delayed via some co-routine method.

File: src/screencast_keys/ops.py (hunk 10 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1547]: Changed.

File: src/screencast_keys/ops.py (hunk 11 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1546]: Changed.

File: src/screencast_keys/ops.py (hunk 12 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1561]: Changed, if you mean between 1560 and 1561 lines.

File: src/screencast_keys/preferences.py (hunk 6 above)
Comment (CheeryLee) [reply src/screencast_keys/preferences.py:413]: Turned this back and moved `load_on_startup` to UI section.

File: src/screencast_keys/preferences.py (hunk 5 above)
Comment (CheeryLee) [reply src/screencast_keys/preferences.py:259]: Moved `load_on_startup` to UI section

File: src/screencast_keys/ops.py (hunk 2 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:543]: Changed.

**Comment by CheeryLee** [comment]
Made changes from the list above

File: src/screencast_keys/ops.py (hunk 12 above)
Comment (nutti) [reply src/screencast_keys/ops.py:1552]: Sorry for the unclear comment.

My suggestion is adding the empty line to the last of file.

File: src/screencast_keys/__init__.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (nutti) [reply src/screencast_keys/__init__.py:60]: Understood.

File: src/screencast_keys/ops.py (hunk 7 above)
Comment (nutti) [reply src/screencast_keys/ops.py:1543]: Thanks for the clarification.

Is it possible to add the comment describing here like:

```suggestion
        # If we call bpy.ops.wm.sk_screencast_keys directly, no changes are made because bpy.context.area is not loaded at that moment. 
        # Therefore, we need to call bpy.ops.wm.sk_screencast_keys with delay via co-routine method.
        cls.initialization_handler = bpy.types.SpaceView3D.draw_handler_add(cls.intialization_callback, (None, None), 'WINDOW', 'POST_PIXEL')
```

File: src/screencast_keys/preferences.py [hunk 14]
```diff
@@ -298,6 +298,12 @@ def ui_in_sidebar_update_fn(self, context):
         update=ui_in_sidebar_update_fn,
     )
 
+    load_on_startup = bpy.props.BoolProperty(
+        name="Load On Startup",
+        description="""Automatically enable screencast when blender is starting up""",
```
Comment (nutti) [inline src/screencast_keys/preferences.py:303]:
```suggestion
    enable_on_startup = bpy.props.BoolProperty(
        name="Enable On Startup",
        description="""Automatically enable Screencast Keys when blender is starting up""",
```

File: src/screencast_keys/preferences.py [hunk 15]
```diff
@@ -383,12 +389,14 @@ def draw(self, context):
             if self.show_last_operator:
                 col.prop(self, "last_operator_show_mode")
 
+            layout.separator()
+
+            layout.label(text="UI:")
+            col = layout.column()
+            col.prop(self, "load_on_startup")
```
Comment (nutti) [inline src/screencast_keys/preferences.py:396]: I think this option is not related to UI, so it is appropriate to move to after `layout.separator` located on L356.

**Review by nutti — CHANGES_REQUESTED** [review]
Great works @CheeryLee .

Almost looks good to me.
Could you check my additional comments?
We can merge your PR after applying my request change.

File: src/screencast_keys/ops.py (hunk 12 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1552]: Ok, got it.

File: src/screencast_keys/preferences.py (hunk 14 above)
Comment (CheeryLee) [reply src/screencast_keys/preferences.py:303]: Changed.

File: src/screencast_keys/preferences.py (hunk 15 above)
Comment (CheeryLee) [reply src/screencast_keys/preferences.py:396]: Changed.

**Comment by nutti** [comment]
@CheeryLee 

Thanks for your work.
I think you missed [this comment](https://github.com/nutti/Screencast-Keys/pull/62#discussion_r671238162).
Could you apply this suggestion?
After applying this suggestion, I can merge your patch :)

File: src/screencast_keys/ops.py (hunk 7 above)
Comment (CheeryLee) [reply src/screencast_keys/ops.py:1546]: Yeah, I forgot about that. Changed.

**Review by nutti — APPROVED** [review]
LGTM

Thanks for your hard work @CheeryLee !
